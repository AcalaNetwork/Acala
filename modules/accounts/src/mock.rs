//! Mocks for the cdp treasury module.

#![cfg(test)]

use super::*;
use frame_support::{
	impl_outer_dispatch, impl_outer_event, impl_outer_origin, ord_parameter_types, parameter_types,
	weights::IdentityFee,
};
use frame_system::EnsureSignedBy;
use primitives::Amount;
use sp_core::H256;
use sp_runtime::{testing::Header, traits::IdentityLookup, FixedPointNumber, Perbill};
use sp_std::cell::RefCell;
use support::{CDPTreasury, EmergencyShutdown, Rate, Ratio};

pub type AccountId = u128;
pub type BlockNumber = u64;
pub type Moment = u64;
pub type Share = u64;

pub const ALICE: AccountId = 1;
pub const BOB: AccountId = 2;
pub const CAROL: AccountId = 3;
pub const ACA: CurrencyId = CurrencyId::ACA;
pub const AUSD: CurrencyId = CurrencyId::AUSD;
pub const BTC: CurrencyId = CurrencyId::XBTC;

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Runtime;

impl_outer_origin! {
	pub enum Origin for Runtime {}
}

impl_outer_dispatch! {
	pub enum Call for Runtime where origin: Origin {
		orml_currencies::Currencies,
		frame_system::System,
	}
}

impl_outer_event! {
	pub enum TestEvent for Runtime {
		frame_system<T>,
		orml_tokens<T>,
		pallet_balances<T>,
		orml_currencies<T>,
		dex<T>,
	}
}

parameter_types! {
	pub const BlockHashCount: u64 = 250;
	pub const MaximumBlockWeight: u32 = 1024;
	pub const MaximumBlockLength: u32 = 2 * 1024;
	pub const AvailableBlockRatio: Perbill = Perbill::one();
}

impl frame_system::Trait for Runtime {
	type Origin = Origin;
	type Index = u64;
	type BlockNumber = BlockNumber;
	type Call = Call;
	type Hash = H256;
	type Hashing = ::sp_runtime::traits::BlakeTwo256;
	type AccountId = AccountId;
	type Lookup = IdentityLookup<Self::AccountId>;
	type Header = Header;
	type Event = TestEvent;
	type BlockHashCount = BlockHashCount;
	type MaximumBlockWeight = MaximumBlockWeight;
	type MaximumBlockLength = MaximumBlockLength;
	type AvailableBlockRatio = AvailableBlockRatio;
	type Version = ();
	type ModuleToIndex = ();
	type AccountData = pallet_balances::AccountData<Balance>;
	type OnNewAccount = ();
	type OnKilledAccount = ();
	type DbWeight = ();
	type BlockExecutionWeight = ();
	type ExtrinsicBaseWeight = ();
	type MaximumExtrinsicWeight = ();
	type BaseCallFilter = ();
	type SystemWeightInfo = ();
}
pub type System = frame_system::Module<Runtime>;

impl orml_tokens::Trait for Runtime {
	type Event = TestEvent;
	type Balance = Balance;
	type Amount = Amount;
	type CurrencyId = CurrencyId;
	type OnReceived = Accounts;
	type WeightInfo = ();
}
pub type Tokens = orml_tokens::Module<Runtime>;

parameter_types! {
	pub const ExistentialDeposit: Balance = 0;
}

impl pallet_balances::Trait for Runtime {
	type Balance = Balance;
	type DustRemoval = ();
	type Event = TestEvent;
	type ExistentialDeposit = ExistentialDeposit;
	type AccountStore = Accounts;
	type WeightInfo = ();
}
pub type PalletBalances = pallet_balances::Module<Runtime>;

pub type AdaptedBasicCurrency =
	orml_currencies::BasicCurrencyAdapter<PalletBalances, Balance, Balance, Amount, BlockNumber>;

parameter_types! {
	pub const GetNativeCurrencyId: CurrencyId = ACA;
}

impl orml_currencies::Trait for Runtime {
	type Event = TestEvent;
	type MultiCurrency = Tokens;
	type NativeCurrency = AdaptedBasicCurrency;
	type GetNativeCurrencyId = GetNativeCurrencyId;
	type WeightInfo = ();
}
pub type Currencies = orml_currencies::Module<Runtime>;

parameter_types! {
	pub const MinimumPeriod: Moment = 5;
}

impl pallet_timestamp::Trait for Runtime {
	type Moment = Moment;
	type OnTimestampSet = ();
	type MinimumPeriod = MinimumPeriod;
	type WeightInfo = ();
}
pub type TimeModule = pallet_timestamp::Module<Runtime>;

parameter_types! {
	pub const TransactionByteFee: Balance = 2;
}

impl pallet_transaction_payment::Trait for Runtime {
	type Currency = PalletBalances;
	type OnTransactionPayment = ();
	type TransactionByteFee = TransactionByteFee;
	type WeightToFee = IdentityFee<Balance>;
	type FeeMultiplierUpdate = ();
}

pub struct MockCDPTreasury;
impl CDPTreasury<AccountId> for MockCDPTreasury {
	type Balance = Balance;
	type CurrencyId = CurrencyId;

	fn get_surplus_pool() -> Self::Balance {
		Default::default()
	}
	fn get_debit_pool() -> Self::Balance {
		Default::default()
	}
	fn get_total_collaterals(_: Self::CurrencyId) -> Self::Balance {
		Default::default()
	}
	fn get_debit_proportion(_: Self::Balance) -> Ratio {
		Default::default()
	}
	fn on_system_debit(_: Self::Balance) -> DispatchResult {
		Ok(())
	}
	fn on_system_surplus(_: Self::Balance) -> DispatchResult {
		Ok(())
	}
	fn issue_debit(_: &AccountId, _: Self::Balance, _: bool) -> DispatchResult {
		Ok(())
	}
	fn burn_debit(_: &AccountId, _: Self::Balance) -> DispatchResult {
		Ok(())
	}
	fn deposit_surplus(_: &AccountId, _: Self::Balance) -> DispatchResult {
		Ok(())
	}
	fn deposit_collateral(_: &AccountId, _: Self::CurrencyId, _: Self::Balance) -> DispatchResult {
		Ok(())
	}
	fn withdraw_collateral(_: &AccountId, _: Self::CurrencyId, _: Self::Balance) -> DispatchResult {
		Ok(())
	}
}

thread_local! {
	static IS_SHUTDOWN: RefCell<bool> = RefCell::new(false);
}

pub struct MockEmergencyShutdown;
impl EmergencyShutdown for MockEmergencyShutdown {
	fn is_shutdown() -> bool {
		IS_SHUTDOWN.with(|v| *v.borrow_mut())
	}
}

ord_parameter_types! {
	pub const Zero: AccountId = 0;
}

parameter_types! {
	pub GetExchangeFee: Rate = Rate::saturating_from_rational(0, 100);
	pub const GetStableCurrencyId: CurrencyId = AUSD;
	pub EnabledCurrencyIds: Vec<CurrencyId> = vec![ACA, BTC];
	pub const DEXModuleId: ModuleId = ModuleId(*b"aca/dexm");
}

impl dex::Trait for Runtime {
	type Event = TestEvent;
	type Currency = Currencies;
	type Share = Share;
	type EnabledCurrencyIds = EnabledCurrencyIds;
	type GetBaseCurrencyId = GetStableCurrencyId;
	type GetExchangeFee = GetExchangeFee;
	type CDPTreasury = MockCDPTreasury;
	type UpdateOrigin = EnsureSignedBy<Zero, AccountId>;
	type ModuleId = DEXModuleId;
	type EmergencyShutdown = MockEmergencyShutdown;
}
pub type DEXModule = dex::Module<Runtime>;

parameter_types! {
	pub const FreeTransferCount: u8 = 3;
	pub const FreeTransferPeriod: Moment = 100;
	pub const FreeTransferDeposit: Balance = 200;
	pub AllNonNativeCurrencyIds: Vec<CurrencyId> = vec![AUSD, BTC];
	pub const NewAccountDeposit: Balance = 100;
	pub const TreasuryModuleId: ModuleId = ModuleId(*b"py/trsry");
	pub MaxSlippageSwapWithDEX: Ratio = Ratio::one();
}

impl Trait for Runtime {
	type FreeTransferCount = FreeTransferCount;
	type FreeTransferPeriod = FreeTransferPeriod;
	type Time = TimeModule;
	type FreeTransferDeposit = FreeTransferDeposit;
	type AllNonNativeCurrencyIds = AllNonNativeCurrencyIds;
	type NativeCurrencyId = GetNativeCurrencyId;
	type Currency = Currencies;
	type DEX = DEXModule;
	type OnCreatedAccount = frame_system::CallOnCreatedAccount<Runtime>;
	type KillAccount = frame_system::CallKillAccount<Runtime>;
	type NewAccountDeposit = NewAccountDeposit;
	type TreasuryModuleId = TreasuryModuleId;
	type MaxSlippageSwapWithDEX = MaxSlippageSwapWithDEX;
}
pub type Accounts = Module<Runtime>;

pub struct ExtBuilder {
	endowed_accounts: Vec<(AccountId, CurrencyId, Balance)>,
}

impl Default for ExtBuilder {
	fn default() -> Self {
		Self {
			endowed_accounts: vec![(ALICE, AUSD, 10000), (ALICE, BTC, 1000)],
		}
	}
}

impl ExtBuilder {
	pub fn build(self) -> sp_io::TestExternalities {
		let mut t = frame_system::GenesisConfig::default()
			.build_storage::<Runtime>()
			.unwrap();

		pallet_balances::GenesisConfig::<Runtime> {
			balances: vec![(ALICE, 100000 + NewAccountDeposit::get())],
		}
		.assimilate_storage(&mut t)
		.unwrap();

		orml_tokens::GenesisConfig::<Runtime> {
			endowed_accounts: self.endowed_accounts,
		}
		.assimilate_storage(&mut t)
		.unwrap();

		t.into()
	}
}
