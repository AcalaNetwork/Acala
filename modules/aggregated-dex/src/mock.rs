// This file is part of Acala.

// Copyright (C) 2022 Acala Foundation.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! Mocks for the Aggregated DEX module.

#![cfg(test)]

use super::*;
use frame_support::{
	match_types, ord_parameter_types, parameter_types,
	traits::{ConstU128, ConstU32, ConstU64, Everything, Nothing},
	PalletId,
};
use frame_system::EnsureSignedBy;
use orml_tokens::ConvertBalance;
pub use orml_traits::{parameter_type_with_key, MultiCurrency};
use primitives::{Amount, TokenSymbol, TradingPair};
use sp_runtime::{
	testing::{Header, H256},
	traits::{Bounded, IdentityLookup},
	AccountId32, FixedPointNumber,
};
pub use support::{ExchangeRate, RebasedStableAsset};

pub type AccountId = AccountId32;
pub type BlockNumber = u64;

mod aggregated_dex {
	pub use super::super::*;
}

pub const ALICE: AccountId = AccountId32::new([1u8; 32]);
pub const BOB: AccountId = AccountId32::new([2u8; 32]);
pub const AUSD: CurrencyId = CurrencyId::Token(TokenSymbol::AUSD);
pub const DOT: CurrencyId = CurrencyId::Token(TokenSymbol::DOT);
pub const LDOT: CurrencyId = CurrencyId::Token(TokenSymbol::LDOT);
pub const STABLE_ASSET: CurrencyId = CurrencyId::StableAssetPoolToken(0);

impl frame_system::Config for Runtime {
	type BaseCallFilter = Everything;
	type BlockWeights = ();
	type BlockLength = ();
	type Origin = Origin;
	type Call = Call;
	type Index = u64;
	type BlockNumber = BlockNumber;
	type Hash = H256;
	type Hashing = ::sp_runtime::traits::BlakeTwo256;
	type AccountId = AccountId;
	type Lookup = IdentityLookup<Self::AccountId>;
	type Header = Header;
	type Event = Event;
	type BlockHashCount = ConstU64<250>;
	type DbWeight = ();
	type Version = ();
	type PalletInfo = PalletInfo;
	type AccountData = pallet_balances::AccountData<Balance>;
	type OnNewAccount = ();
	type OnKilledAccount = ();
	type SystemWeightInfo = ();
	type SS58Prefix = ();
	type OnSetCode = ();
	type MaxConsumers = ConstU32<16>;
}

parameter_type_with_key! {
	pub ExistentialDeposits: |_currency_id: CurrencyId| -> Balance {
		Default::default()
	};
}

impl orml_tokens::Config for Runtime {
	type Event = Event;
	type Balance = Balance;
	type Amount = Amount;
	type CurrencyId = CurrencyId;
	type WeightInfo = ();
	type ExistentialDeposits = ExistentialDeposits;
	type OnDust = ();
	type MaxLocks = ();
	type MaxReserves = ();
	type ReserveIdentifier = [u8; 8];
	type DustRemovalWhitelist = Nothing;
	type OnNewTokenAccount = ();
	type OnKilledTokenAccount = ();
}

ord_parameter_types! {
	pub const Admin: AccountId = BOB;
}

parameter_types! {
	pub const DEXPalletId: PalletId = PalletId(*b"aca/dexm");
	pub const GetExchangeFee: (u32, u32) = (0, 100);
	pub EnabledTradingPairs: Vec<TradingPair> = vec![];
}

impl module_dex::Config for Runtime {
	type Event = Event;
	type Currency = Tokens;
	type GetExchangeFee = GetExchangeFee;
	type TradingPathLimit = ConstU32<4>;
	type PalletId = DEXPalletId;
	type Erc20InfoMapping = ();
	type DEXIncentives = ();
	type WeightInfo = ();
	type ListingOrigin = EnsureSignedBy<Admin, AccountId>;
	type ExtendedProvisioningBlocks = ConstU64<0>;
	type OnLiquidityPoolUpdated = ();
}

pub struct EnsurePoolAssetId;
impl nutsfinance_stable_asset::traits::ValidateAssetId<CurrencyId> for EnsurePoolAssetId {
	fn validate(currency_id: CurrencyId) -> bool {
		matches!(currency_id, CurrencyId::StableAssetPoolToken(_))
	}
}

pub struct ConvertBalanceHoma;
impl ConvertBalance<Balance, Balance> for ConvertBalanceHoma {
	type AssetId = CurrencyId;

	fn convert_balance(balance: Balance, asset_id: CurrencyId) -> Balance {
		match asset_id {
			LDOT => ExchangeRate::saturating_from_rational(1, 10)
				.checked_mul_int(balance)
				.unwrap_or(Bounded::max_value()),
			_ => balance,
		}
	}

	fn convert_balance_back(balance: Balance, asset_id: CurrencyId) -> Balance {
		match asset_id {
			LDOT => ExchangeRate::saturating_from_rational(10, 1)
				.checked_mul_int(balance)
				.unwrap_or(Bounded::max_value()),
			_ => balance,
		}
	}
}

match_types! {
	pub type IsLiquidToken: impl Contains<CurrencyId> = {
		CurrencyId::Token(TokenSymbol::LDOT)
	};
}

type RebaseTokens = orml_tokens::Combiner<
	AccountId,
	IsLiquidToken,
	orml_tokens::Mapper<AccountId, Tokens, ConvertBalanceHoma, Balance, GetLiquidCurrencyId>,
	Tokens,
>;

parameter_types! {
	pub const StableAssetPalletId: PalletId = PalletId(*b"nuts/sta");
}

impl nutsfinance_stable_asset::Config for Runtime {
	type Event = Event;
	type AssetId = CurrencyId;
	type Balance = Balance;
	type Assets = RebaseTokens;
	type PalletId = StableAssetPalletId;

	type AtLeast64BitUnsigned = u128;
	type FeePrecision = ConstU128<10_000_000_000>; // 10 decimals
	type APrecision = ConstU128<100>; // 2 decimals
	type PoolAssetLimit = ConstU32<5>;
	type SwapExactOverAmount = ConstU128<100>;
	type WeightInfo = ();
	type ListingOrigin = EnsureSignedBy<Admin, AccountId>;
	type EnsurePoolAssetId = EnsurePoolAssetId;
}

parameter_types! {
	pub static DexSwapJointList: Vec<Vec<CurrencyId>> = vec![];
	pub const GetLiquidCurrencyId: CurrencyId = LDOT;
}

impl Config for Runtime {
	type DEX = Dex;
	type StableAsset = StableAssetWrapper;
	type GovernanceOrigin = EnsureSignedBy<Admin, AccountId>;
	type DexSwapJointList = DexSwapJointList;
	type SwapPathLimit = ConstU32<3>;
	type WeightInfo = ();
}

pub type StableAssetWrapper =
	RebasedStableAsset<StableAsset, ConvertBalanceHoma, RebasedStableAssetErrorConvertor<Runtime>>;

type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Runtime>;
type Block = frame_system::mocking::MockBlock<Runtime>;

frame_support::construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
		AggregatedDex: aggregated_dex::{Pallet, Call, Storage},
		Dex: module_dex::{Pallet, Call, Storage, Config<T>, Event<T>},
		Tokens: orml_tokens::{Pallet, Storage, Event<T>, Config<T>},
		StableAsset: nutsfinance_stable_asset::{Pallet, Call, Storage, Event<T>},
	}
);

pub struct ExtBuilder {
	balances: Vec<(AccountId, CurrencyId, Balance)>,
}

impl Default for ExtBuilder {
	fn default() -> Self {
		Self {
			balances: vec![
				(ALICE, DOT, 100_000_000_000),
				(BOB, AUSD, 1_000_000_000_000_000_000),
				(BOB, DOT, 1_000_000_000_000_000_000),
				(BOB, LDOT, 1_000_000_000_000_000_000),
			],
		}
	}
}

impl ExtBuilder {
	pub fn build(self) -> sp_io::TestExternalities {
		let mut t = frame_system::GenesisConfig::default()
			.build_storage::<Runtime>()
			.unwrap();

		orml_tokens::GenesisConfig::<Runtime> {
			balances: self.balances,
		}
		.assimilate_storage(&mut t)
		.unwrap();

		t.into()
	}
}
