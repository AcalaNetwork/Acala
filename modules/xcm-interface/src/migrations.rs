// This file is part of Acala.

// Copyright (C) 2020-2022 Acala Foundation.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

// This file is used for initial migration from HomaXcm into XcmInterface, due to name change.
use frame_support::{
	traits::{Get, GetStorageVersion, PalletInfoAccess, StorageVersion},
	weights::Weight,
};

pub mod v1 {
	use super::*;
	use crate::*;

	/// Migrate the entire storage of previously named "module-homa-xcm" pallet to here.
	pub fn migrate<T: frame_system::Config, P: GetStorageVersion + PalletInfoAccess>() -> Weight {
		let old_prefix = "HomaXcm";
		let new_prefix = "XcmInterface";

		let on_chain_storage_version = <P as GetStorageVersion>::on_chain_storage_version();

		log::info!(
			target: "runtime::xcm-interface",
			"Running migration from HomaXcm to XcmInterface. \n
			Old prefix: {:?}, New prefix: {:?} \n
			Current version: {:?}, New version: 1",
			old_prefix, new_prefix, on_chain_storage_version,
		);

		if on_chain_storage_version < 1 {
			frame_support::storage::migration::move_pallet(old_prefix.as_bytes(), new_prefix.as_bytes());
			StorageVersion::new(1).put::<P>();
			log::info!(
				target: "runtime::xcm-interface",
				"Storage migrated from HomaXcm to XcmInterface.",
			);
			<T as frame_system::Config>::BlockWeights::get().max_block
		} else {
			log::warn!(
				target: "runtime::xcm-interface",
				"Attempted to apply migration to v1 but failed because storage version is {:?}",
				on_chain_storage_version,
			);
			0
		}
	}
}
