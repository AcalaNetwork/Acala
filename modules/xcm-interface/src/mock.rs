// This file is part of Acala.

// Copyright (C) 2020-2023 Acala Foundation.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! Mocks for the prices module.

#![cfg(test)]

use super::*;
use crate as xcm_interface;
use frame_support::{
	construct_runtime, ord_parameter_types, parameter_types,
	traits::{ConstU128, ConstU32, ConstU64, Everything, Nothing},
};
use frame_system::EnsureSignedBy;
use primitives::{CurrencyId, TokenSymbol};
use sp_core::H256;
use sp_runtime::{testing::Header, traits::IdentityLookup, AccountId32};
use xcm_builder::{EnsureXcmOrigin, FixedWeightBounds, SignedToAccountId32};

pub type AccountId = AccountId32;
pub type BlockNumber = u64;

pub const ALICE: AccountId = AccountId32::new([1u8; 32]);
pub const DOT: CurrencyId = CurrencyId::Token(TokenSymbol::DOT);

impl frame_system::Config for Runtime {
	type RuntimeOrigin = RuntimeOrigin;
	type Index = u64;
	type BlockNumber = BlockNumber;
	type RuntimeCall = RuntimeCall;
	type Hash = H256;
	type Hashing = ::sp_runtime::traits::BlakeTwo256;
	type AccountId = AccountId;
	type Lookup = IdentityLookup<Self::AccountId>;
	type Header = Header;
	type RuntimeEvent = RuntimeEvent;
	type BlockHashCount = ConstU64<250>;
	type BlockWeights = ();
	type BlockLength = ();
	type Version = ();
	type PalletInfo = PalletInfo;
	type AccountData = pallet_balances::AccountData<Balance>;
	type OnNewAccount = ();
	type OnKilledAccount = ();
	type DbWeight = ();
	type BaseCallFilter = Everything;
	type SystemWeightInfo = ();
	type SS58Prefix = ();
	type OnSetCode = ();
	type MaxConsumers = ConstU32<16>;
}

impl pallet_balances::Config for Runtime {
	type Balance = Balance;
	type DustRemoval = ();
	type RuntimeEvent = RuntimeEvent;
	type ExistentialDeposit = ConstU128<1>;
	type AccountStore = System;
	type MaxLocks = ();
	type MaxReserves = ();
	type ReserveIdentifier = [u8; 8];
	type WeightInfo = ();
}

parameter_types! {
	pub const UnitWeightCost: XcmWeight = XcmWeight::from_parts(10, 10);
	pub const BaseXcmWeight: XcmWeight = XcmWeight::from_parts(100_000_000, 100_000_000);
	pub const MaxInstructions: u32 = 100;
	pub const MaxAssetsIntoHolding: u32 = 64;
}

parameter_types! {
	pub const RelayNetwork: NetworkId = NetworkId::Polkadot;
	pub UniversalLocation: InteriorMultiLocation =
		X1(Parachain(2000).into());
}

pub type LocalOriginToLocation = SignedToAccountId32<RuntimeOrigin, AccountId, RelayNetwork>;

#[cfg(feature = "runtime-benchmarks")]
parameter_types! {
	pub ReachableDest: Option<MultiLocation> = Some(Parent.into());
}

impl pallet_xcm::Config for Runtime {
	type RuntimeEvent = RuntimeEvent;
	type SendXcmOrigin = EnsureXcmOrigin<RuntimeOrigin, LocalOriginToLocation>;
	type XcmRouter = ();
	type ExecuteXcmOrigin = EnsureXcmOrigin<RuntimeOrigin, LocalOriginToLocation>;
	type XcmExecuteFilter = Everything;
	type XcmExecutor = ();
	type XcmTeleportFilter = Nothing;
	type XcmReserveTransferFilter = Everything;
	type Weigher = FixedWeightBounds<UnitWeightCost, RuntimeCall, MaxInstructions>;
	type UniversalLocation = UniversalLocation;
	type RuntimeOrigin = RuntimeOrigin;
	type RuntimeCall = RuntimeCall;
	const VERSION_DISCOVERY_QUEUE_SIZE: u32 = 100;
	type AdvertisedXcmVersion = pallet_xcm::CurrentXcmVersion;
	type Currency = Balances;
	type CurrencyMatcher = ();
	type TrustedLockers = ();
	type SovereignAccountOf = ();
	type MaxLockers = ConstU32<8>;
	type WeightInfo = pallet_xcm::TestWeightInfo;
	#[cfg(feature = "runtime-benchmarks")]
	type ReachableDest = ReachableDest;
}

ord_parameter_types! {
	pub const One: AccountId = ALICE;
}

parameter_types! {
	pub const GetStakingCurrencyId: CurrencyId = DOT;
	pub const ParachainAccount: AccountId = AccountId32::new([0u8; 32]);
	pub const ParachainId: module_relaychain::ParaId = module_relaychain::ParaId::new(2000);
}

pub struct SubAccountIndexMultiLocationConvertor;
impl Convert<u16, MultiLocation> for SubAccountIndexMultiLocationConvertor {
	fn convert(_sub_account_index: u16) -> MultiLocation {
		(Parent, Parachain(2000)).into()
	}
}

pub struct MockXcmTransfer;
impl XcmTransfer<AccountId, Balance, CurrencyId> for MockXcmTransfer {
	fn transfer(
		_who: AccountId,
		_currency_id: CurrencyId,
		_amount: Balance,
		_dest: MultiLocation,
		_dest_weight_limit: WeightLimit,
	) -> DispatchResult {
		unimplemented!()
	}

	/// Transfer `MultiAsset`
	fn transfer_multi_asset(
		_who: AccountId,
		_asset: MultiAsset,
		_dest: MultiLocation,
		_dest_weight_limit: WeightLimit,
	) -> DispatchResult {
		unimplemented!()
	}

	/// Transfer `MultiAssetWithFee`
	fn transfer_multiasset_with_fee(
		_who: AccountId,
		_asset: MultiAsset,
		_fee: MultiAsset,
		_dest: MultiLocation,
		_dest_weight_limit: WeightLimit,
	) -> DispatchResult {
		unimplemented!()
	}
}

impl Config for Runtime {
	type RuntimeEvent = RuntimeEvent;
	type UpdateOrigin = EnsureSignedBy<One, AccountId>;
	type StakingCurrencyId = GetStakingCurrencyId;
	type ParachainAccount = ParachainAccount;
	type RelayChainUnbondingSlashingSpans = ConstU32<28>;
	type SovereignSubAccountLocationConvert = SubAccountIndexMultiLocationConvertor;
	type RelayChainCallBuilder = module_relaychain::RelayChainCallBuilder<Runtime, ParachainId>;
	type XcmTransfer = MockXcmTransfer;
}

type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Runtime>;
type Block = frame_system::mocking::MockBlock<Runtime>;

construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
		Balances: pallet_balances::{Pallet, Call, Storage, Event<T>},
		PolkadotXcm: pallet_xcm::{Pallet, Call, Event<T>, Origin},
		XcmInterface: xcm_interface::{Pallet, Storage, Call, Event<T>},
	}
);

pub struct ExtBuilder;

impl Default for ExtBuilder {
	fn default() -> Self {
		ExtBuilder
	}
}

impl ExtBuilder {
	pub fn build(self) -> sp_io::TestExternalities {
		let t = frame_system::GenesisConfig::default()
			.build_storage::<Runtime>()
			.unwrap();

		let mut ext = sp_io::TestExternalities::new(t);
		ext.execute_with(|| System::set_block_number(1));
		ext
	}
}
