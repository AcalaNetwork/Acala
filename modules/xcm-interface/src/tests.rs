// This file is part of Acala.

// Copyright (C) 2020-2023 Acala Foundation.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! Unit tests for xcm interface module.

#![cfg(test)]

use super::*;
use crate::migrations::{MigrateXcmDestWeightAndFee, OldMultiLocation, OldXcmInterfaceOperation, OldXcmWeight};
use crate::mock::{ExtBuilder, Runtime};
use frame_support::{
	storage::migration::{get_storage_value, put_storage_value},
	traits::OnRuntimeUpgrade,
	StorageHasher, StoragePrefixedMap,
};

#[test]
fn simulate_migrate_xcm_dest_weight_and_fee() {
	ExtBuilder::default().build().execute_with(|| {
		let module_prefix = XcmDestWeightAndFee::<Runtime>::module_prefix();
		let storage_prefix = XcmDestWeightAndFee::<Runtime>::storage_prefix();

		let old_key_1: OldXcmInterfaceOperation = OldXcmInterfaceOperation::XtokensTransfer;
		let old_value_1: (OldXcmWeight, Balance) = (1_000_000_000, 200_000_000);
		let old_key_2: OldXcmInterfaceOperation = OldXcmInterfaceOperation::ParachainFee(Box::new(
			OldMultiLocation::new(1, xcm::v2::Junctions::X1(xcm::v2::Junction::Parachain(1000))),
		));
		let old_value_2: (OldXcmWeight, Balance) = (2_000_000_000, 500_000_000);
		let new_key_1: XcmInterfaceOperation = XcmInterfaceOperation::XtokensTransfer;
		let new_value_1: (XcmWeight, Balance) = (XcmWeight::from_parts(1_000_000_000, 1024 * 128), 200_000_000);
		let new_key_2: XcmInterfaceOperation =
			XcmInterfaceOperation::ParachainFee(Box::new(MultiLocation::new(1, X1(Parachain(1000)))));
		let new_value_2: (XcmWeight, Balance) = (XcmWeight::from_parts(2_000_000_000, 1024 * 128), 500_000_000);

		// put old raw storage
		put_storage_value(
			module_prefix,
			storage_prefix,
			&Twox64Concat::hash(&old_key_1.encode()),
			old_value_1,
		);
		put_storage_value(
			module_prefix,
			storage_prefix,
			&Twox64Concat::hash(&old_key_2.encode()),
			old_value_2,
		);
		assert_eq!(
			get_storage_value::<(OldXcmWeight, Balance)>(
				module_prefix,
				storage_prefix,
				&Twox64Concat::hash(&old_key_1.encode()),
			),
			Some(old_value_1)
		);
		assert_eq!(
			get_storage_value::<(OldXcmWeight, Balance)>(
				module_prefix,
				storage_prefix,
				&Twox64Concat::hash(&old_key_2.encode()),
			),
			Some(old_value_2)
		);

		// Run migration
		assert_eq!(
			MigrateXcmDestWeightAndFee::<Runtime>::on_runtime_upgrade(),
			<<Runtime as frame_system::Config>::DbWeight as Get<frame_support::weights::RuntimeDbWeight>>::get()
				.reads_writes(2, 2)
		);
		assert_eq!(
			get_storage_value::<(XcmWeight, Balance)>(
				module_prefix,
				storage_prefix,
				&Twox64Concat::hash(&new_key_1.encode()),
			),
			Some(new_value_1)
		);
		assert_eq!(
			get_storage_value::<(XcmWeight, Balance)>(
				module_prefix,
				storage_prefix,
				&Twox64Concat::hash(&new_key_2.encode()),
			),
			Some(new_value_2)
		);
	});
}

// TODO: other unit tests
