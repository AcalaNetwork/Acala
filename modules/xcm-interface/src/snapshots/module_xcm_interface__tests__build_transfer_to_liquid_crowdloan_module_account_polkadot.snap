---
source: modules/xcm-interface/src/tests.rs
expression: xcm
---
Xcm(
    [
        WithdrawAsset(
            MultiAssets(
                [
                    MultiAsset {
                        id: Concrete(
                            MultiLocation {
                                parents: 0,
                                interior: Here,
                            },
                        ),
                        fun: Fungible(
                            0,
                        ),
                    },
                ],
            ),
        ),
        BuyExecution {
            fees: MultiAsset {
                id: Concrete(
                    MultiLocation {
                        parents: 0,
                        interior: Here,
                    },
                ),
                fun: Fungible(
                    0,
                ),
            },
            weight_limit: Unlimited,
        },
        Transact {
            origin_kind: SovereignAccount,
            require_weight_at_most: Weight {
                ref_time: 0,
                proof_size: 0,
            },
            call: "0x1d0000010101010101010101010101010101010101010101010101010101010101010100630803000100411f03000101000202020202020202020202020202020202020202020202020202020202020202030400000000070010a5d4e80000000000",
        },
        RefundSurplus,
        DepositAsset {
            assets: Wild(
                AllCounted(
                    1,
                ),
            ),
            beneficiary: MultiLocation {
                parents: 0,
                interior: X1(
                    Parachain(
                        2000,
                    ),
                ),
            },
        },
    ],
)
